README.md
# SpaceInv
Old school Space Invaders game

[Imgur](https://i.imgur.com/bfqgYUf.png)


Backup/Changelog.txt
0.0
- Alpha version with geometric shapes

0.1
- Beta version with images and sounds
- Original release based on Christian Thompson's creation

0.2
- Score record implemented
- Rows of enemies implemented
- Level change implemented
- Image path fixes
- Sound path fixes
- Minor bug fixes

0.3
- Enemy speed increase implemented
- Score record improvement
- Screen size change
- Minor bug fixes

0.4
- Life implemented
- Pause implemented
- Weapon speed tweak

0.5
- Enemy fire implemented
- Level cycle implemented
- Overall image improvement
- Enemy speed increase tweak
- Play area tweak
- First row bug fixed
- PEP 8 Style applied

0.6
- Score record improvement
- Overall image improvement
- Minor code tweak

1.0 - Release

subs.py
import random

key = {'1':'a', '2':'i', '3':'l', '4':'u',
    '5':'j', '6':'r', '7':'o', '8':'t', '9':'v'}
door = 'bcdefghkmnpqswxyz'

def decrypt(text):
    """Decrypt string to number"""
    text = text[7:11] + text[27:]
    decripted = ""
    for char in text:
        if char in door:
            decripted += "0"
        else:
            s = [n for n, c in key.items() if c == char]
            decripted += str(s[0])
    return decripted

def encrypt(text):
    """Encrypt number to string"""
    text = str(text)
    add = ""
    for _ in range(7 - len(text)):
        text = "0" + text
    for char in text:
        if char == "0":
            add += random.choice(door)
        else:
            add += key[char]
    encripted = ""
    for _ in range(7):
        encripted += chr(random.randrange(97, 123))
    encripted += add[:4]
    for _ in range(16):
        encripted += chr(random.randrange(97, 123))
    encripted += add[4:]
    return encripted


Backup/main_00.py
import os
import turtle

# Screen
screen = turtle.Screen()
screen.title("Space Invaders")
screen.bgcolor("black")
screen.setup(width = 800, height = 700)
screen.tracer(0)

# Border
border_pen = turtle.Turtle()
border_pen.speed(0)
border_pen.hideturtle()
border_pen.color("white")
border_pen.pensize(3)
border_pen.penup()
border_pen.goto(-300, -300)
border_pen.pendown()
for side in range(4):
    border_pen.fd(600)
    border_pen.lt(90)

# Enemy
enemy = turtle.Turtle()
enemy.speed(0)
enemy.shape("circle")
enemy.color("yellow")
enemy.penup()
enemy.goto(-200, 240)
x_speed = 0.2
y_speed = 40

# Player
player = turtle.Turtle()
player.speed(0)
player.shape("triangle")
player.color("lime")
player.penup()
player.goto(0, -250)
player.setheading(90)
playerspeed = 15

# Player weapon
weapon = turtle.Turtle()
weapon.speed(0)
weapon.hideturtle()
weapon.shape("arrow")
weapon.shapesize(0.1, 0.5)
weapon.color("white")
weapon.penup()
weapon.setheading(90)
weaponspeed = 2
weaponstate = 0
weapon.goto(0, -400)

# Fire weapon
def fire_weapon():
    # Declare weaponstate as a global if it needs changes
    global weaponstate
    if weaponstate == 0:
        weaponstate = 1
        # Move the missile to just above the player
        x = player.xcor()
        y = player.ycor() + 10
        weapon.goto(x, y)

# Check hit
def is_hit(t1, t2):
    distance = (
        (t1.xcor() - t2.xcor()) ** 2 + (t1.ycor() - t2.ycor()) ** 2
    ) ** 0.5
    return True if distance < 15 else False

# Move player
def move_right():
    x = player.xcor()
    x += playerspeed
    if x > 280:
        x = 280
    player.setx(x)

def move_left():
    x = player.xcor()
    x -= playerspeed
    if x < -280:
        x = -280
    player.setx(x)

# Keyboard binding
screen.listen()
screen.onkeypress(move_right, "Right")
screen.onkeypress(move_left, "Left")
screen.onkeypress(fire_weapon, "space")

# Main game loop
while True:
    screen.update()
    
    # Move enemy
    x_e = enemy.xcor()
    x_e += x_speed
    enemy.setx(x_e)
    if x_e > 280:
        y_e = enemy.ycor()
        y_e -= y_speed
        x_speed *= -1
        enemy.sety(y_e)
    elif x_e < -280:
        y_e = enemy.ycor()
        y_e -= y_speed
        x_speed *= -1
        enemy.sety(y_e)
    
    # Move missile
    if weaponstate == 1:
        weapon.showturtle()
        y_w = weapon.ycor()
        y_w += weaponspeed
        weapon.sety(y_w)
    if weapon.ycor() > 280:
        weapon.hideturtle()
        weaponstate = 0
        weapon.goto(0, -400)
    
    # Check for a hit
    if is_hit(weapon, enemy):
        weapon.hideturtle()
        weaponstate = 0
        weapon.goto(0, -400)
        enemy.goto(-200, 240)

    # Check for enemy reaching player
    if is_hit(player, enemy):
        player.hideturtle()
        enemy.hideturtle()
        print('Game Over')
        break

input('Press enter to finish')


Backup/main_01.py
import os
import random
import turtle
import winsound

# Screen
screen = turtle.Screen()
screen.title("Space Invaders")
screen.bgcolor("black")
screen.setup(width = 800, height = 700)
screen.tracer(0)

# Register shapes
screen.register_shape("_player.gif")
screen.register_shape("_enemy_1.gif")
screen.register_shape("_enemy_2.gif")
screen.register_shape("_enemy_3.gif")

# Border
border_pen = turtle.Turtle()
border_pen.speed(0)
border_pen.hideturtle()
border_pen.color("white")
border_pen.pensize(3)
border_pen.penup()
border_pen.goto(-300, -300)
border_pen.pendown()
for side in range(4):
    border_pen.fd(600)
    border_pen.lt(90)

# Score
score_pen = turtle.Turtle()
score_pen.speed(0)
score_pen.hideturtle()
score_pen.color("white")
score_pen.penup()
score_pen.goto(-290, 280)
score = 0
score_str = f"Score: {score}"
score_pen.write(score_str, False, align="left", font=("Courier", 12, "normal"))

# Enemies
num_enemies = 10
enemies = []
for i in range(num_enemies):
    enemies.append(turtle.Turtle())
for enemy in enemies:
    enemy.speed(0)
    enemy.shape("_enemy_1.gif")
    enemy.color("white")
    enemy.penup()
    x = random.randint(-200, 200)
    y = random.randint(120, 240)
    enemy.goto(x, y)
    # enemy.goto(-200, 240)     # original enemy position
x_speed = 0.1
y_speed = 40

# Player
player = turtle.Turtle()
player.speed(0)
player.shape("_player.gif")
player.color("lime")
player.penup()
player.goto(0, -250)
player.setheading(90)
playerspeed = 15

# Player weapon
weapon = turtle.Turtle()
weapon.speed(0)
weapon.hideturtle()
weapon.shape("arrow")
weapon.shapesize(0.1, 0.5)
weapon.color("white")
weapon.penup()
weapon.goto(0, -400)
weapon.setheading(90)
weaponspeed = 2
weaponstate = 0

# Fire weapon
def fire_weapon():
    # Declare weaponstate as a global if it needs changes
    global weaponstate
    if weaponstate == 0:
        winsound.PlaySound("s_laser.wav", winsound.SND_ASYNC)
        weaponstate = 1
        # Move the missile to just above the player
        x = player.xcor()
        y = player.ycor() + 10
        weapon.goto(x, y)

# Check hit
def is_hit(t1, t2):
    distance = (
        (t1.xcor() - t2.xcor()) ** 2 + (t1.ycor() - t2.ycor()) ** 2
    ) ** 0.5
    return True if distance < 15 else False

# Move player
def move_right():
    x = player.xcor()
    x += playerspeed
    if x > 280:
        x = 280
    player.setx(x)

def move_left():
    x = player.xcor()
    x -= playerspeed
    if x < -280:
        x = -280
    player.setx(x)

# Keyboard binding
screen.listen()
screen.onkeypress(move_right, "Right")
screen.onkeypress(move_left, "Left")
screen.onkeypress(fire_weapon, "space")

# Main game loop
gameover = False
while not gameover:
    screen.update()

    # Move enemies
    for enemy in enemies:
        x_e = enemy.xcor()
        x_e += x_speed
        enemy.setx(x_e)
        if x_e > 280:
            x_speed *= -1
            for e in enemies:
                y_e = e.ycor()
                y_e -= y_speed
                e.sety(y_e)
        elif x_e < -280:
            x_speed *= -1
            for e in enemies:
                y_e = e.ycor()
                y_e -= y_speed
                e.sety(y_e)

        # Check for a hit
        if is_hit(weapon, enemy):
            winsound.PlaySound("s_explod.wav", winsound.SND_ASYNC)
            weapon.hideturtle()
            weaponstate = 0
            weapon.goto(0, -400)
            x = random.randint(-200, 200)
            y = random.randint(120, 240)
            enemy.goto(x, y)
            # Update score
            score += 10
            score_str = f"Score: {score}"
            score_pen.clear()
            score_pen.write(score_str, False, align="left", font=("Courier", 12, "normal"))

        # Check for enemy reaching player
        if is_hit(player, enemy) or enemy.ycor() <= player.ycor():
            winsound.PlaySound("s_gm_ovr.wav", winsound.SND_ASYNC)
            screen.update()
            print('Game Over')
            gameover = True

    # Move missile
    if weaponstate == 1:
        weapon.showturtle()
        y_w = weapon.ycor()
        y_w += weaponspeed
        weapon.sety(y_w)
    if weapon.ycor() > 280:
        weapon.hideturtle()
        weaponstate = 0
        weapon.goto(0, -400)

input('Press enter to finish')


Backup/main_02.py
# TODO speed increase if wall bump
# TODO enemy fire

import os
import random
import time
import turtle
import winsound

# Screen
screen = turtle.Screen()
screen.title("Space Invaders")
screen.bgcolor("black")
screen.setup(width = 800, height = 700)
screen.tracer(0)

# Register shapes
screen.register_shape(".\Images\_player.gif")
screen.register_shape(".\Images\_enemy_1.gif")
screen.register_shape(".\Images\_enemy_2.gif")
screen.register_shape(".\Images\_enemy_3.gif")

# Border
border_pen = turtle.Turtle()
border_pen.speed(0)
border_pen.hideturtle()
border_pen.color("white")
border_pen.pensize(3)
border_pen.penup()
border_pen.goto(-300, -300)
border_pen.pendown()
for side in range(4):
    border_pen.fd(600)
    border_pen.lt(90)

# Disclaimer
disc_pen = turtle.Turtle()
disc_pen.speed(0)
disc_pen.hideturtle()
disc_pen.color("white")
disc_pen.penup()
disc_pen.goto(0, 0)

# Score
with open ('score.ttl', "r") as f:
    record = int(f.read())
score_pen = turtle.Turtle()
score_pen.speed(0)
score_pen.hideturtle()
score_pen.color("white")
score_pen.penup()
score_pen.goto(-290, 260)
score = 0
score_str = f"Record: {record}\nScore: {score}"
score_pen.write(score_str, False, align="left", font=("Courier", 12, "normal"))

# Player
player = turtle.Turtle()
player.speed(0)
player.shape(".\Images\_player.gif")
player.penup()
player.goto(0, -250)
playerspeed = 15

# Player weapon
weapon = turtle.Turtle()
weapon.speed(0)
weapon.hideturtle()
weapon.shape("arrow")
weapon.shapesize(0.1, 0.5)
weapon.color("white")
weapon.penup()
weapon.goto(0, -400)
weapon.setheading(90)
weaponspeed = 2
weaponstate = 0

# Fire weapon
def fire_weapon():
    # Declare weaponstate as a global if it needs changes
    global weaponstate
    if weaponstate == 0:
        winsound.PlaySound(".\Sounds\s_laser.wav", winsound.SND_ASYNC)
        weaponstate = 1
        # Move the missile to just above the player
        x = player.xcor()
        y = player.ycor() + 10
        weapon.goto(x, y)

# Check hit
def is_hit(t1, t2):
    distance = (
        (t1.xcor() - t2.xcor()) ** 2 + (t1.ycor() - t2.ycor()) ** 2
    ) ** 0.5
    return True if distance < 15 else False

# Move player
def move_right():
    x = player.xcor()
    x += playerspeed
    if x > 280:
        x = 280
    player.setx(x)

def move_left():
    x = player.xcor()
    x -= playerspeed
    if x < -280:
        x = -280
    player.setx(x)

# Keyboard binding
screen.listen()
screen.onkeypress(move_right, "Right")
screen.onkeypress(move_left, "Left")
screen.onkeypress(fire_weapon, "space")

# Main game loop
level = 1
gameover = False
while not gameover:
    screen.update()

    # Enemies
    enemies_rows = 5
    enemies_in_row = 11
    enemies = [[], [], [], [], []]
    x_origin = -220
    y_origin = 280 - 40 * level
    y_space = 0
    for i in range(enemies_rows):
        for j in range(enemies_in_row):
            enemies[i].append(turtle.Turtle())
        x_space = 0
        for enemy in enemies[i]:
            if y_space == 0:
                enemy.shape(".\Images\_enemy_3.gif")
            elif y_space < -80:
                enemy.shape(".\Images\_enemy_1.gif")
            else:
                enemy.shape(".\Images\_enemy_2.gif")
            enemy.speed(0)
            enemy.penup()
            enemy.goto(x_origin + x_space, y_origin + y_space)
            x_space += 40
        y_space -= 40
    x_speed = 0.1
    y_speed = 40
    e_alive = enemies_rows * enemies_in_row
    
    disc_pen.write(f"Level {level}", False, align="center", font=("Courier", 28, "normal"))
    time.sleep(3)
    disc_pen.clear()

    next_lvl = False
    while not next_lvl:
        screen.update()
        
        # Move enemies
        for i in range(enemies_rows):
            temp_lis = enemies[0].copy()
            for enemy in enemies[i]:
                enemy.setx(enemy.xcor() + x_speed)

                # Check for a hit
                if is_hit(weapon, enemy):
                    winsound.PlaySound(".\Sounds\s_explod.wav", winsound.SND_ASYNC)
                    weapon.hideturtle()
                    weaponstate = 0
                    weapon.goto(0, -400)
                    e_alive -= 1
                    if i == 0:
                        temp_lis = [k for k in enemies[0] if k != enemy]
                    enemy.goto(0, 1000)
                    # Update score
                    if i < 1:
                        score += 30
                    elif i < 3:
                        score += 20
                    else:
                        score += 10
                    score_str = f"Record: {record}\nScore: {score}"
                    score_pen.clear()
                    score_pen.write(score_str, False, align="left", font=("Courier", 12, "normal"))

                # Check for enemy reaching player
                if is_hit(player, enemy) or enemy.ycor() <= player.ycor():
                    winsound.PlaySound(".\Sounds\s_gm_ovr.wav", winsound.SND_ASYNC)
                    screen.update()
                    disc_pen.write(
                        f"GAME OVER", False, align="center", font=("Courier", 28, "normal")
                        )
                    print('Game Over')
                    if score > record:
                        time.sleep(3)
                        disc_pen.clear()
                        disc_pen.write(
                            f"New Record: {score}", False, align="center", font=("Courier", 28, "normal")
                            )                        
                        with open ('score.ttl', "w") as f:
                            f.write(str(score))
                    gameover = True
                    next_lvl = True
            enemies[0] = temp_lis.copy()
        
        # Level & Wall bump
        if e_alive == 0:
            # Next level
            print('Next Level')
            level += 1
            next_lvl = True
        else:
            # Check wall bump
            if enemies[0][0].xcor() > 280 or enemies[0][len(enemies[0])-1].xcor() > 280:
                x_speed *= -1
                for i in range(enemies_rows):
                    for enemy in enemies[i]:
                        enemy.sety(enemy.ycor() - y_speed)
            elif enemies[0][0].xcor() < -280 or enemies[0][len(enemies[0])-1].xcor() < -280:
                x_speed *= -1
                for i in range(enemies_rows):
                    for enemy in enemies[i]:
                        enemy.sety(enemy.ycor() - y_speed)

        # Move missile
        if weaponstate == 1:
            weapon.showturtle()
            y_w = weapon.ycor()
            y_w += weaponspeed
            weapon.sety(y_w)
        if weapon.ycor() > 280:
            weapon.hideturtle()
            weaponstate = 0
            weapon.goto(0, -400)

time.sleep(3)


Backup/main_06.py
"""
Space invaders by Pedro Tortello - 24/01/2019
"""

import random
import subs
import time
import turtle
import winsound

# Enemy fire
def enemy_fire(enemy):
    global e_weaponstate
    e_weaponstate = 1
    # Move the missile to just below the enemy
    x = enemy.xcor()
    y = enemy.ycor() - 10
    e_weapon.goto(x, y)

# Fire weapon
def fire_weapon():
    # Declare weaponstate as a global if it needs changes
    global weaponstate
    if weaponstate == 0:
        winsound.PlaySound(".\Sounds\s_laser.wav", winsound.SND_ASYNC)
        weaponstate = 1
        # Move the missile to just above the player
        x = player.xcor()
        y = player.ycor() + 10
        weapon.goto(x, y)

# Check hit
def is_hit(t1, t2):
    distance = (
        (t1.xcor() - t2.xcor()) ** 2 + (t1.ycor() - t2.ycor()) ** 2
    ) ** 0.5
    return True if distance < 15 else False

# Check wall bump
def check_wall():
    global x_speed
    for i in range(enemies_rows):
        for enemy in enemies[i]:
            if enemy.ycor() < 400:
                if enemy.xcor() > 230 or enemy.xcor() < -230:
                    x_speed *= -1
                    for i in range(enemies_rows):
                        for enemy in enemies[i]:
                            enemy.sety(enemy.ycor() - y_speed)
                    return

# Life lost
def lifelost(trtl):
    global lifes
    global level
    global e_alive
    lifes -= 1
    level -= 1
    e_alive = 0
    for i in range(2):
        time.sleep(0.5)
        trtl.hideturtle()
        screen.update()
        time.sleep(0.5)
        trtl.showturtle()
        screen.update()
    time.sleep(1)
    trtl.hideturtle()
    screen.update()    

# Move player
def move_right():
    x = player.xcor()
    x += playerspeed
    if x > 230:
        x = 230
    player.setx(x)

def move_left():
    x = player.xcor()
    x -= playerspeed
    if x < -230:
        x = -230
    player.setx(x)

def pause_game():
    global gameplay
    if gameplay:
        gameplay = False
        player.hideturtle()
        weapon.hideturtle()
        e_weapon.hideturtle()
        for i in range(enemies_rows):
            for enemy in enemies[i]:
                enemy.hideturtle()
        disc_pen.write(f"Game paused", False, align="center",
                        font=("Courier", 28, "normal"))
        while not gameplay:
            screen.onkeypress(pause_game, "Right")
            screen.onkeypress(pause_game, "Left")
            screen.onkeypress(pause_game, "space")
            screen.update()
    else:
        gameplay = True
        time.sleep(0.5)
        disc_pen.clear()
        player.showturtle()
        weapon.showturtle()
        e_weapon.showturtle()
        for i in range(enemies_rows):
            for enemy in enemies[i]:
                enemy.showturtle()
        screen.onkeypress(move_right, "Right")
        screen.onkeypress(move_left, "Left")
        screen.onkeypress(fire_weapon, "space")

# Screen set up
screen = turtle.Screen()
screen.title("Space Invaders")
screen.setup(width = 700, height = 800)

# Register shapes
screen.register_shape(".\Images\_player.gif")
screen.register_shape(".\Images\_destroy.gif")
screen.register_shape(".\Images\_enemy_1.gif")
screen.register_shape(".\Images\_enemy_2.gif")
screen.register_shape(".\Images\_enemy_3.gif")
screen.register_shape(".\Images\_e_laser.gif")

# Main game loop
gameplay = True     # Set not pause condition
new_game = True
level = 1
lifes = 2
gameover = False
while not gameover:

    # Screen
    screen.bgcolor("black")
    screen.tracer(0)

    # Border
    border_pen = turtle.Turtle()
    border_pen.speed(0)
    border_pen.hideturtle()
    border_pen.color("white")
    border_pen.pensize(3)
    border_pen.penup()
    border_pen.goto(-250, -350)
    border_pen.pendown()
    v = -100
    for side in range(4):
        border_pen.fd(600 + v)
        border_pen.lt(90)
        v *= -1

    # Disclaimer
    disc_pen = turtle.Turtle()
    disc_pen.speed(0)
    disc_pen.hideturtle()
    disc_pen.color("white")
    disc_pen.penup()
    disc_pen.goto(0, 0)

    # Disclaimer background
    disc_bg = turtle.Turtle()
    disc_bg.speed(0)
    disc_bg.shape("square")
    disc_bg.shapesize(stretch_wid=2, stretch_len=22)
    disc_bg.hideturtle()
    disc_bg.color("black")
    disc_bg.penup()
    disc_bg.goto(0, 20)

    # Score
    score_pen = turtle.Turtle()
    score_pen.speed(0)
    score_pen.hideturtle()
    score_pen.color("white")
    score_pen.penup()
    score_pen.goto(0, 350)
    scorecor = []
    with open ('subs.ttl', "r") as f:
        for line in f:
            scorecor.append(line.strip())
        scorecor[0] = int(scorecor[0])
        scorecor[1] = int(subs.decrypt(scorecor[1]))
    if new_game:
        new_game = False
        scorecor[0] = 0
        with open ('subs.ttl', "w") as f:
            f.write(str(scorecor[0]) + "\n" + subs.encrypt(scorecor[1]))
    score_str = f"Score: {scorecor[0]}\t\t\tRecord: {scorecor[1]}"
    score_pen.write(score_str, False, align="center",
                    font=("Courier", 12, "normal"))

    # Player
    player = turtle.Turtle()
    player.speed(0)
    player.shape(".\Images\_player.gif")
    player.penup()
    player.goto(0, -300)
    playerspeed = 15

    # Player lifes
    life1 = turtle.Turtle()
    life1.speed(0)
    life1.shape(".\Images\_player.gif")
    life1.penup()
    life1.goto(-230, -330)
    if lifes < 1:
        life1.hideturtle()
    life2 = turtle.Turtle()
    life2.speed(0)
    life2.shape(".\Images\_player.gif")
    life2.penup()
    life2.goto(-200, -330)
    if lifes < 2:
        life2.hideturtle()

    # Player weapon
    weapon = turtle.Turtle()
    weapon.speed(0)
    weapon.hideturtle()
    weapon.shape("arrow")
    weapon.shapesize(0.1, 0.5)
    weapon.color("white")
    weapon.penup()
    weapon.goto(0, -450)
    weapon.setheading(90)
    weaponspeed = 5
    weaponstate = 0

    # Keyboard binding
    screen.listen()
    screen.onkeypress(move_right, "Right")
    screen.onkeypress(move_left, "Left")
    screen.onkeypress(fire_weapon, "space")
    screen.onkeypress(pause_game, "P")
    screen.onkeypress(pause_game, "p")

    screen.update()

    # Enemies
    enemies_rows = 5
    enemies_in_row = 11
    enemies = [[], [], [], [], []]
    x_origin = -220
    mod_level = level % 10
    if mod_level == 0:
        mod_level = 10
    y_origin = 340 - 40 * mod_level
    y_space = 0
    for i in range(enemies_rows):
        for j in range(enemies_in_row):
            enemies[i].append(turtle.Turtle())
        x_space = 0
        for enemy in enemies[i]:
            if y_space == 0:
                enemy.shape(".\Images\_enemy_3.gif")
            elif y_space < -80:
                enemy.shape(".\Images\_enemy_1.gif")
            else:
                enemy.shape(".\Images\_enemy_2.gif")
            enemy.speed(0)
            enemy.penup()
            enemy.goto(x_origin + x_space, y_origin + y_space)
            x_space += 40
        y_space -= 40
    speed_in = 0
    x_speed = 0.1
    y_speed = 40
    e_alive = enemies_rows * enemies_in_row
    
    # Enemy weapon
    e_weapon = turtle.Turtle()
    e_weapon.speed(0)
    e_weapon.hideturtle()
    e_weapon.shape(".\Images\_e_laser.gif")
    e_weapon.penup()
    e_weapon.goto(0, 450)
    e_weapon.setheading(270)
    e_weaponspeed = 1
    e_weaponstate = 0

    # Level loop
    disc_pen.write(f"Level {level}", False, align="center",
                    font=("Courier", 28, "normal"))
    time.sleep(3)
    disc_pen.clear()
    next_lvl = False
    while not next_lvl:
        screen.update()
        
        # Move enemies
        for i in range(enemies_rows):
            for enemy in enemies[i]:
                enemy.setx(enemy.xcor() + x_speed)

                # Check for a hit
                if is_hit(weapon, enemy):
                    winsound.PlaySound(".\Sounds\s_explod.wav",
                                        winsound.SND_ASYNC)
                    weapon.hideturtle()
                    weaponstate = 0
                    weapon.goto(0, -450)
                    e_alive -= 1
                    speed_in += 1
                    enemy.goto(0, 1500)
                    # Update score
                    if i < 1:
                        scorecor[0] += 30
                    elif i < 3:
                        scorecor[0] += 20
                    else:
                        scorecor[0] += 10
                    score_str = (
                        f"Score: {scorecor[0]}\t\t\tRecord: {scorecor[1]}"
                    )
                    score_pen.clear()
                    score_pen.write(score_str, False, align="center",
                                    font=("Courier", 12, "normal"))

                # Check for life loss
                if is_hit(player, enemy) or enemy.ycor() <= player.ycor() or (
                    is_hit(player, e_weapon)):
                    winsound.PlaySound(
                        ".\Sounds\s_life.wav", winsound.SND_ASYNC)
                    player.shape(".\Images\_destroy.gif")
                    e_weapon.hideturtle()
                    e_weaponstate = 0
                    e_weapon.goto(0, 450)
                    screen.update()
                    if lifes == 2:
                        lifelost(life2)
                    elif lifes == 1:
                        lifelost(life1)
                    else:
                        disc_bg.showturtle()
                        screen.update()
                        disc_pen.write(f"GAME OVER", False, align="center",
                                        font=("Courier", 28, "normal"))
                        time.sleep(3)
                        disc_pen.clear()
                        if scorecor[0] > scorecor[1]:
                            disc_pen.write(
                                f"New Record: {scorecor[0]}", False,
                                align="center", font=("Courier", 28, "normal")
                                )
                            time.sleep(3)
                            with open ('subs.ttl', "w") as f:
                                f.write(
                                    str(0) + "\n" + subs.encrypt(scorecor[0]))
                        else:
                            with open ('subs.ttl', "w") as f:
                                f.write(
                                    str(0) + "\n" + subs.encrypt(scorecor[1]))
                        gameover = True
                        next_lvl = True
                    break

        # Speed increase
        if speed_in == 7:
            speed_in = 0
            if x_speed > 0:
                x_speed += 0.01
            else:
                x_speed -= 0.01
        elif e_alive < 7:
            if x_speed > 0:
                x_speed = 0.2 + (6 - e_alive) * 0.04
            else:
                x_speed = -0.2 - (6 - e_alive) * 0.04

        # Level & Wall bump
        if e_alive == 0:
            # Next level
            level += 1
            next_lvl = True
            with open ('subs.ttl', "w") as f:
                f.write(str(scorecor[0]) + "\n" + subs.encrypt(scorecor[1]))
            turtle.clearscreen()
            for i in range(enemies_rows):
                for enemy in enemies[i]:
                    del enemy
        else:
            # Check for wall bump
            check_wall()

        # Move missile
        if weaponstate == 1:
            weapon.showturtle()
            weapon.sety(weapon.ycor() + weaponspeed)
        if weapon.ycor() > 340:
            weapon.hideturtle()
            weaponstate = 0
            weapon.goto(0, -450)

        # Move enemy missile
        if e_weaponstate == 1:
            e_weapon.showturtle()
            e_weapon.sety(e_weapon.ycor() - e_weaponspeed)
        else:
            # Enemy firing
            row_to_col = list(zip(*enemies))
            armed = []
            for i in range(len(row_to_col)):
                lower = [400, 0]
                for enemy in row_to_col[i]:
                    if enemy.ycor() < lower[0]:
                        lower = [enemy.ycor(), enemy]
                if lower[0] < 400:
                    armed.append(lower[1])
            enemy_fire(random.choice(armed))
        if e_weapon.ycor() < -340:
            e_weapon.hideturtle()
            e_weaponstate = 0
            e_weapon.goto(0, 450)
        
        # Check for weapon vs weapon hit
        if is_hit(weapon, e_weapon):
            weapon.hideturtle()
            weaponstate = 0
            weapon.goto(0, -450)
            e_weapon.hideturtle()
            e_weaponstate = 0
            e_weapon.goto(0, 450)
